server/app/controllers/applicant/RedirectController.java

```diff
@@ -69,37 +72,74 @@ public CompletionStage<Result> programByName(Http.Request request, String progra
               // the information controller to ask for preferred language.
               if (!applicant.getApplicantData().hasPreferredLocale()) {
-                return redirect(
-                        controllers.applicant.routes.ApplicantInformationController.edit(
-                            applicant.id))
-                    .withSession(request.session().adding(REDIRECT_TO_SESSION_KEY, request.uri()));
+                return CompletableFuture.completedFuture(
+                    redirect(
+                            controllers.applicant.routes.ApplicantInformationController.edit(
+                                applicantId))
+                        .withSession(
+                            request.session().adding(REDIRECT_TO_SESSION_KEY, request.uri())));
               }
 
-              return redirect(
-                  controllers.applicant.routes.ApplicantProgramReviewController.preview(
-                      applicant.id, programFuture.join().id));
+              return getProgramVersionForApplicant(applicantId, programSlug)
+                  .thenComposeAsync(
+                      (Optional<ProgramDefinition> programForExistingApplication) -> {
+                        // Check to see if the applicant already has an application
+                        // for this program, redirect to program version associated
+                        // with that application if so.
+                        if (programForExistingApplication.isPresent()) {
+                          return CompletableFuture.completedFuture(
+                              redirect(
+                                  controllers.applicant.routes.ApplicantProgramReviewController
+                                      .preview(
+                                          applicantId, programForExistingApplication.get().id())));
+                        }
+
+                        return redirectToActiveProgram(applicantId, programSlug);
+                      },
+                      httpContext.current());
             },
             httpContext.current());
   }
 
+  private CompletionStage<Result> redirectToActiveProgram(long applicantId, String programSlug) {
+    return programService
+        .getActiveProgramDefinitionAsync(programSlug)
+        .thenApplyAsync(
+            (activeProgramDefinition) ->
+                redirect(
+                    controllers.applicant.routes.ApplicantProgramReviewController.preview(
+                        applicantId, activeProgramDefinition.id())),
+            httpContext.current());
+  }
+
+  private CompletionStage<Optional<ProgramDefinition>> getProgramVersionForApplicant(
+      long applicantId, String programSlug) {
+    // Find all applicant's DRAFT applications for programs of the same slug
+    // redirect to the newest program version with a DRAFT application.
+    // TODO(#2573): clean this up, don't compare on program ID
```

Comment:
how's #2573 relevant here? you're not comparing on program ids.

Reply:
We are on line 138.

Brett's PR for #2573 has query code that will make this less janky and allow us to compare on the actual applications